recursion/core/src/poseidon2_wide/external.rs

```diff
@@ -292,48 +285,56 @@ fn eval_external_round<AB: SP1AirBuilder>(
 
 fn eval_internal_rounds<AB: SP1AirBuilder>(
     builder: &mut AB,
-    cols: &Poseidon2WideCols<AB::Var>,
+    cols: &Poseidon2ColType<AB::Var>,
     is_real: AB::Var,
 ) {
-    let round_cols = &cols.internal_rounds;
-    let mut state: [AB::Expr; WIDTH] = core::array::from_fn(|i| round_cols.state[i].into());
+    let poseidon2_cols = cols.get_poseidon2_cols();
+    let state = &poseidon2_cols.internal_rounds_state;
+    let s0 = poseidon2_cols.internal_rounds_s0;
+    let sbox_3 = cols.get_internal_sbox();
+    let mut state: [AB::Expr; WIDTH] = core::array::from_fn(|i| state[i].into());
     for r in 0..NUM_INTERNAL_ROUNDS {
         // Add the round constant.
         let round = r + NUM_EXTERNAL_ROUNDS / 2;
         let add_rc = if r == 0 {
             state[0].clone()
         } else {
-            round_cols.s0[r - 1].into()
+            s0[r - 1].into()
         } + is_real * AB::Expr::from_wrapped_u32(RC_16_30_U32[round][0]);
 
-        let sbox_deg_3 = add_rc.clone() * add_rc.clone() * add_rc.clone();
-        builder.assert_eq(round_cols.sbox_deg_3[r], sbox_deg_3);
+        let mut sbox_deg_3 = add_rc.clone() * add_rc.clone() * add_rc.clone();
+        if let Some(expected) = sbox_3 {
+            builder.assert_eq(expected[r], sbox_deg_3);
+            sbox_deg_3 = expected[r].into();
+        }
 
         // See `populate_internal_rounds` for why we don't have columns for the sbox output here.
-        let sbox_deg_7 =
-            round_cols.sbox_deg_3[r].into() * round_cols.sbox_deg_3[r].into() * add_rc.clone();
+        let sbox_deg_7 = sbox_deg_3.clone() * sbox_deg_3 * add_rc.clone();
 
         // Apply the linear layer.
         // See `populate_internal_rounds` for why we don't have columns for the new state here.
         state[0] = sbox_deg_7.clone();
         internal_linear_layer(&mut state);
 
         if r < NUM_INTERNAL_ROUNDS - 1 {
-            builder.assert_eq(round_cols.s0[r], state[0].clone());
+            builder.assert_eq(s0[r], state[0].clone());
         }
     }
 
+    let external_state = poseidon2_cols.external_rounds_state[NUM_EXTERNAL_ROUNDS / 2];
     for i in 0..WIDTH {
-        builder.assert_eq(
-            cols.external_rounds[NUM_EXTERNAL_ROUNDS / 2].state[i],
-            state[i].clone(),
-        )
+        builder.assert_eq(external_state[i], state[i].clone())
     }
 }
 
 impl<F, const DEGREE: usize> BaseAir<F> for Poseidon2WideChip<DEGREE> {
     fn width(&self) -> usize {
-        NUM_POSEIDON2_WIDE_COLS
+        let use_sbox_3 = DEGREE < 7;
```

Comment:
I think it would be nicer to use a match statement (although the compiler might already optimize this if away)

```
match DEGREE {
    d if d < 7 => NUM_POSEIDON2_SBOX_COLS
    _ => NUM_POSEIDON2_COLS
}
```